rotkehlchen/chain/ethereum/modules/uniswap/v3/decoder.py

```diff
@@ -57,93 +68,237 @@ def _maybe_decode_v3_swap(  # pylint: disable=no-self-use
+
+        if (decoded_events[1].asset == A_ETH and decoded_events[1].counterparty in UNISWAP_ROUTERS):  # noqa: E501
+            if decoded_events[1].event_type == HistoryEventType.SPEND:
+                # This is a swap from eth to token
+                swap_data = self._decode_eth_to_token_swap(decoded_events)
+
+            elif decoded_events[1].event_type == HistoryEventType.RECEIVE:
+                # This is a swap from token to eth
+                swap_data = self._decode_token_to_eth_swap(decoded_events)
+
+        else:
+            # Then this should be a swap from token to token
+            swap_data = self._decode_token_to_token_swap(decoded_events)
+
+        if swap_data.from_asset is None or swap_data.to_asset is None:
+            self.msg_aggregator.add_warning(
+                f'Failed to decode a uniswap swap for transaction {transaction.tx_hash.hex()}. '
+                f'Couldnt find either `from` or `to` asset',
+            )
+            return decoded_events
+
+        try:
+            from_crypto_asset = swap_data.from_asset.resolve_to_crypto_asset()
+        except (UnknownAsset, WrongAssetType):
+            self.notify_user(event=decoded_events[1], counterparty=CPT_UNISWAP_V3)
+            return decoded_events
+
+        try:
+            to_crypto_asset = swap_data.to_asset.resolve_to_crypto_asset()
+        except (UnknownAsset, WrongAssetType):
+            self.notify_user(event=decoded_events[-1], counterparty=CPT_UNISWAP_V3)
+            return decoded_events
+
+        from_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=decoded_events[1].sequence_index,
+            timestamp=decoded_events[1].timestamp,
+            location=Location.BLOCKCHAIN,
+            location_label=transaction.from_address,
+            asset=from_crypto_asset,
+            balance=Balance(amount=swap_data.from_amount),
+            notes=f'Swap {swap_data.from_amount} {from_crypto_asset.symbol} via uniswap auto router',  # noqa: E501
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.SPEND,
+            counterparty=CPT_UNISWAP_V3,
+        )
+
+        to_event = HistoryBaseEntry(
+            event_identifier=transaction.tx_hash,
+            sequence_index=decoded_events[-1].sequence_index,
+            timestamp=decoded_events[-1].timestamp,
+            location=Location.BLOCKCHAIN,
+            location_label=transaction.from_address,
+            asset=to_crypto_asset,
+            balance=Balance(amount=swap_data.to_amount),
+            notes=f'Receive {swap_data.to_amount} {to_crypto_asset.symbol} as the result of a swap via uniswap auto router',  # noqa: E501
+            event_type=HistoryEventType.TRADE,
+            event_subtype=HistoryEventSubType.RECEIVE,
+            counterparty=CPT_UNISWAP_V3,
+        )
+
+        gas_event = decoded_events[0]
```

Comment:
should you check that this is indeed the gas event?

Reply:
Could make sense to make sure that the transaction was initiated by the user :+1: